src/contrib/mlir/compiler.cpp

```diff
@@ -301,131 +303,136 @@ namespace ngraph
-                return compiler.create_unary_op<mlir::NGReluOp>(ng_node);
+                auto ng_node_gather = static_cast<const ngraph::op::Gather*>(ng_node);
+                mlir::Value* result = compiler.create_generic_op<mlir::NGGatherOp>(ng_node);
+                mlir::Operation* op = result->getDefiningOp();
+                op->setAttr("axis", compiler.m_builder->getI64IntegerAttr(ng_node_gather->get_axis()));
+                return result;
+
             }
         }
     }
 }
 
-const MLIRCompiler::MLIRCompOpMap MLIRCompiler::op_dispatcher{
-#define MLIR_OP(OP) {TI(ngraph::op::OP), &MLIRCompiler::create_op<ngraph::op::OP>},
-#include "ops_supported.inc"
-};
-
-template <typename UnaryOp>
-mlir::Value* MLIRCompiler::create_unary_op(const ngraph::Node* ng_node)
-{
-    auto lhs = ng_node->get_argument(0)->get_output_tensor_ptr();
-    auto lhs_v = get_tensor_value(lhs.get()).m_value;
-    auto res_type = get_mlir_type(ng_node->get_output_tensor_ptr().get());
-    return m_builder->create<UnaryOp>(mlir::UnknownLoc::get(&m_context), res_type, lhs_v)
-        .getResult();
-}
-
-template <typename BinOp>
-mlir::Value* MLIRCompiler::create_binary_op(const ngraph::Node* ng_node)
-{
-    auto lhs = ng_node->get_argument(0)->get_output_tensor_ptr();
-    auto rhs = ng_node->get_argument(1)->get_output_tensor_ptr();
-    auto lhs_v = get_tensor_value(lhs.get()).m_value;
-    auto rhs_v = get_tensor_value(rhs.get()).m_value;
-    auto res_type = get_mlir_type(ng_node->get_output_tensor_ptr().get());
-    return m_builder->create<BinOp>(mlir::UnknownLoc::get(&m_context), res_type, lhs_v, rhs_v)
-        .getResult();
-}
-
-mlir::Value* MLIRCompiler::create_concat(const ngraph::Node* ng_node)
+template <typename Op>
+mlir::Value* MLIRCompiler::create_generic_op(const ngraph::Node* ng_node)
 {
     std::vector<mlir::Value*> arg_values;
-    auto ng_node_concat = static_cast<const ngraph::op::Concat*>(ng_node);
+    std::vector<mlir::Type> res_types;
     for (auto& arg : ng_node->get_arguments())
     {
         auto arg_tensor = arg->get_output_tensor_ptr();
         auto arg_v = get_tensor_value(arg_tensor.get()).m_value;
         arg_values.push_back(arg_v);
     }
-    auto res_type = get_mlir_type(ng_node->get_output_tensor_ptr().get());
+
+    for (auto& output : ng_node->outputs())
+    {
+        res_types.push_back(get_mlir_type(output.get_tensor_ptr().get()));
+    }
+
     return m_builder
-        ->create<mlir::NGConcatOp>(
-            mlir::UnknownLoc::get(&m_context),
-            res_type,
-            arg_values,
-            m_builder->getI64IntegerAttr(ng_node_concat->get_concatenation_axis()))
-        .getResult();
+        ->create<Op,
+                 ArrayRef<mlir::Type>,
+                 ArrayRef<mlir::Value *>,
+                 ArrayRef<mlir::NamedAttribute>>(
+                    mlir::UnknownLoc::get(&m_context),
+                    res_types,
+                    arg_values, {/* no attrs */}).getResult();
```

Comment:
I think `create_generic_op` should return an op so that we can use it for ops returning multiple values. And the in client, we can get the op's single return or multiple returns.

Reply:
I also changed `create_op` to return Operation instead and handle multi-value def at the dispatcher.